### DISCUSSION TITLE
Github private repositories

### DISCUSSION THREAD

**Discussion by raczosala** [vercel/community#1395] [Help] [answered]
I want to use a private repository, pull the source and then use `next-transpile-modules`.

I've read the [Knowledge Base](https://vercel.com/knowledge/using-private-dependencies-with-vercel) but I'm not sure this is the same use case since my library is not packed, I want to pull the entire source and let `next-transpile-modules` do the transpilation.

I know we can use https and oauth or ssh but I'm not sure how to achieve it with Vercel without the tokens exposed in the repository 

`package-name": "git+https://<github_token>:x-oauth-basic@github.com/<user>/<repo>.git`

is it possible to use Vercel Environment Variables in the package.json?

**Comment by tenax002** [comment]
Any updates on this?

**Comment by amyegan** [comment, marked as answer]
Hi! We have a guide on [using private dependencies with Vercel](https://vercel.com/guides/using-private-dependencies-with-vercel) that should help with this. You can create an .npmrc with the necessary registry info and upload it as an environment variable in Vercel to keep the token out of your repository. More info here: https://vercel.com/guides/using-private-dependencies-with-vercel#other-package-registries

**Reply by nvti** [reply]
Hi @amyegan, follow [this guide](https://vercel.com/guides/using-private-dependencies-with-vercel), I can add a package from a private github repo like: `"package-name": "git+https://<username>:<github_token>@github.com/<user>/<repo>.git"`

But how can I hide `github_token` by using the environment variable? In this guide, Vercel team said: "You can keep access tokens private by storing them as Environment Variables in your project's settings." but they don't have any instruction.

**Reply by amyegan** [reply]
The last section in the guide has some instructions about how to store store your `.npmrc` as an environment variable to give your project access to private packages without uploading the file to your repo: https://vercel.com/guides/using-private-dependencies-with-vercel#other-package-registries

The environment variable documentation has more general information on how to manage them in your project settings: https://vercel.com/docs/concepts/projects/environment-variables

**Reply by kristojorg** [reply]
This guide shows how to set an auth token for the github package registry, but not how to set an environment variable to install a github repo directly. Is that possible?

**Comment by tenax002** [comment]
Hey, thanks for the reply.
It is still quite unclear on how we can reference the private git repo from
package.json once we include the .npmrc file in the root directory as well
as the env variable in vercel.

Can you show a more explicit example?

**Reply by amyegan** [reply]
GitHub has a document describing GitHub packages installation in more detail: [Installing from GitHub Packages Registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-npm-registry#installing-a-package). 

>To add your package dependencies to the package.json file for GitHub Packages, specify the full-scoped package name, such as @my-org/server. For packages from npmjs.com, specify the full name, such as @babel/core.

Vercel will automatically use the `NPM_RC` env var .npmrc you provide for the build environment. Aside from uploading as an environment variable, this works like a normal .npmrc

**Reply by oztune** [reply]
What about the case when you're not using the npm registry and just installing a private repo directly from Github, like:
```
"package": "github:org/package"
```

The document you referenced mentions doing this:

```
"package-name": "git+https://<github_token>:x-oauth-basic@github.com/<user>/<repo>.git"
```

But I'd like to avoid hard-coding the Github access token inside of `package.json`, and provide it an an env variable instead. Is that possible?

It would be great if I could just add an env variable called `GITHUB_ACCESS_TOKEN` or something because committing my personal access token to my code repo seems like an anti-pattern.

**Reply by tenax002** [reply]
You could look into using a pre install script that finds and replaces the `GITHUB_TOKEN` with the actual token from the environment variable. You could combine that with a pre commit hook that ensures the token is reverted back to `GITHUB_TOKEN` instead of the actual token before you commit and push your code.

**Reply by undefined-olha** [reply]
hey @tenax002! Have you actually succeed with that?

**Comment by cvruitenbeekcobbler** [comment]
Currently running into this issue with `yarn` + github workflows/actions. @tenax002 have you succeeded with the NPM set-up?
Might be able to use the same theory for yarn.

Docs I used to attempt getting this to work:
- https://vercel.com/guides/using-private-dependencies-with-vercel
- https://github.com/actions/setup-node/blob/main/docs/advanced-usage.md#use-private-packages
- https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-nodejs#example-using-yarn
- https://yarnpkg.com/cli/config/set (to create a .yarnrc.yml file on the fly)

There's always an "unauthorized" error, while the same token works when directly put in the "environment variables" on Vercel itself.
I also tried to set up a github workflow with NPM, but it results in the same error.

And as @oztune says: I'd rather avoid adding a token into the `package.json`.

If someone could point me to the right direction, that would be great!
